Tests for SysTimer (the idle loop timer for tickless mode)

fkjagodzinski opened the pull request:
## Description

Add tests for an `RtosTimer` class introduced with **TICKLESS mode** support in #4991.
* extracted a `RtosTimer` class from https://github.com/ARMmbed/mbed-os/blob/a519b8449b391cc61d230e0b360e5af936e80ac6/rtos/TARGET_CORTEX/mbed_rtx_idle.cpp#L49 as `rtos/SysTimer`,
* added unit tests.

## Status

**READY**

Issues to be resolved before merge:
* ~#5284~

fkjagodzinski commented:
Regarding CI test runs -- this test suite needs first to be compiled with `-DMBED_TICKLESS` i.e.:

```bash
mbed test -t GCC_ARM -m NUCLEO_L476RG -n tests-mbedmicro-rtos-mbed-systimer -DMBED_TICKLESS --compile
```

and then run without `-DMBED_TICKLESS` i.e.:

```bash
mbed test -t GCC_ARM -m NUCLEO_L476RG -n tests-mbedmicro-rtos-mbed-systimer --run -v
```

Where do I need to include this setup?

bulislaw commented:
`MBED_TICKLESS` should be defined by targets that support tickless feature so I don't think you need to do anything special, it'll be tested by default for supported hw.

fkjagodzinski commented:
@0xc0170 This particular PR adds tests for a class named `RtosTimer` that was introduced with #4991 in https://github.com/ARMmbed/mbed-os/blob/a519b8449b391cc61d230e0b360e5af936e80ac6/rtos/TARGET_CORTEX/mbed_rtx_idle.cpp#L49

> Previously it was internal, now it is being exposed as public API.

This is only to make this class accessible for test code.

> How does this addition stands against RTOSTimer that is deprecated

Apart from the class name -- it's unrelated.

> This is bringing new Timer to the user space.

True. Maybe this could be avoided by putting class files some other place. Could anyone point the appropriate directory?

theotherjimmy commented:
I added a minor release version as minor releases may contain additions to the public set of API's. Patch releases may not contain additions to the public API set.

theotherjimmy commented:
I think this is waiting on clarification/review from @0xc0170

bulislaw commented:
Ah! I've missed the directory change. @0xc0170 good catch! Shall we move it to `rtos/TARGET_CORTEX/` to separate it from RTOS APIs and in doxygen make explicit note that it's not part of the Mbed RTOS API and it's internal implementation of System Timer.

0xc0170 commented:
> Ah! I've missed the directory change. @0xc0170 good catch! Shall we move it to rtos/TARGET_CORTEX/ to separate it from RTOS APIs and in doxygen make explicit note that it's not part of the Mbed RTOS API and it's internal implementation of System Timer.

@fkjagodzinski What do you think? This would work (having it not part of public API, no inclusion besides the file that is using this internal API) and stating in the docs that this is internal detail.

fkjagodzinski commented:
Thanks for the comments! Here is the update:
* moved `SysTimer` files to `rtos/TARGET_CORTEX/`,
* updated doxygen -- now the docs for `SysTimer` will generate only if https://github.com/ARMmbed/mbed-os/blob/4e222952d757e378d5a591ca7fd143dcda24fc7e/doxyfile_options#L637-L641 contains `RTOS_INTERNAL`. *This is my initial prposition regarding doxygen. Do you have any suggestions on how this should be handled?*

bulislaw commented on rtos/TARGET_CORTEX/mbed_rtx_idle.cpp:
```diff
@@ -37,113 +37,18 @@ using namespace mbed;
     // Do not use SingletonPtr since this relies on the RTOS
     if (NULL == os_timer) {
-        os_timer = new (os_timer_data) RtosTimer();
+        os_timer = new (os_timer_data) rtos::SysTimer();
```
Why did we split the constructor into 2 calls?

fkjagodzinski replied:
I did that to avoid manipulating `SysTick` / `IRQ` registers while running tests. I also extracted `set_irq_pending()` from `handler()` for same reason.

0xc0170 commented on rtos/TARGET_CORTEX/SysTimer.h:
```diff
@@ -0,0 +1,115 @@
+#include "platform/NonCopyable.h"
+#include "drivers/TimerEvent.h"
+
+namespace rtos {
```
Looking at the test code, and here. This is in the rtos namespace. Is enough just docs note ?

I would propose to create own  namespace for this within rtos, to show the intent, internal namespace within rtos: `rtos::internal`

fkjagodzinski replied:
Ok, I added an additional `internal` namespace as suggested.

0xc0170 commented:
/morph build

bulislaw commented:
@fkjagodzinski these tests should only be enabled for platforms defining TICKLESS macro.

fkjagodzinski commented:
Update:
* as suggested, tests will run only for targets with `MBED_TICKLESS` defined,
* rebased onto master,
* squashed commits for readability.

Regarding test failures on NRF51_DK -- #5284 has to be resolved first.

0xc0170 commented:
/morph build

fkjagodzinski commented:
Regarding latest CI failures:
* the exporter failure looks unrelated to this PR;
* test failures: 
  * **NUCLEO_F401RE** (`Ticker`, `Timer`, `Timeout`) also seem unrelated; this target doesn't even have `MBED_TICKLESS` defined,
  * **NRF52_DK** & **NRF51_DK** fail on `Schedule zero ticks` case -- this will be fixed with #5284.

0xc0170 commented:
@fkjagodzinski Still then blocked on 5284, I'll restart at least exporters

/morph export-build

cmonr commented:
@fkjagodzinski Has there been any progress made on this PR?

fkjagodzinski commented:
> @fkjagodzinski Has there been any progress made on this PR?

Nope, still blocked by #5284. It seems that Nordic guys are busy with some other things; I'll probably investigate this issue myself when I'm done with my current task.

0xc0170 commented:
@fkjagodzinski Can we do any progress here (is it just blocked on one device?) ?

fkjagodzinski commented:
> [...] is it just blocked on one device?

Yes, NRF51_DK and NRF52_DK fail on `Schedule zero ticks` case, but unfortunately these are the only boards that have TICKLESS enabled.

Regarding the #5284 -- due to other tasks I haven't made any progress yet, but still have that on my plate.

cmonr commented:
@fkjagodzinski We were wondering if https://github.com/ARMmbed/mbed-os/issues/5284 was still on your plate, and this still a precursor to this PR.

jamesbeyond commented:
@cmonr Yes, #5248 and #6162 are precursors to this one.
Maciej was working on #6162, hopefully,  #5248 will be merged soon as well

cmonr commented:
@jamesbeyond @fkjagodzinski Is there an active PR for https://github.com/ARMmbed/mbed-os/issues/5248, since 5248 is currently a closed issue?

0xc0170 commented:
> @jamesbeyond @fkjagodzinski Is there an active PR for #5248, since 5248 is currently a closed issue?

That should be resolved (as stated lab issue).

/morph build

fkjagodzinski commented:
Just to be clear:
* PR #5248 is unrelated,
* issue #5284 is now fixed by PR #5046,
* the reason of the latest morph failure are issues present in test code (I'm investigating those now).

fkjagodzinski commented:
Update:
* rebased onto master, which is now at 8b6a7aa,
* updated tests:
  * removed `test_reschedule()` which duplicated `test_update_tick()`,
  * fixed `test_update_tick()`,
  * fixed `test_deepsleep()`.

@c1728p9, @bulislaw I can see that rebase deprecated your reviews; the `SysTimer` class wasn't modified.

fkjagodzinski commented:
@cmonr, @0xc0170 Looks like this PR is ready. Could you trigger morph tests please?

cmonr commented:
/morph build

cmonr commented:
/morph build

cmonr commented:
/morph build

cmonr commented:
@fkjagodzinski Sorry about the multiple builds. Stale web pages don't auto-refresh, making me think I still need to invoke commands.
